File: DetectCycleInLinkedList.java
// Time Complexity :O(n) length of list
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this :


// Your code here along with comments explaining your approach
1. detect if cycle exists or not by two poinyers fast and slow
2. if exists then set one of the pointer to head and start moving by one again towards end until fast not == slow
3. otherwise return null

/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode detectCycle(ListNode head) {
        
        ListNode dummy = new ListNode(-1);
        dummy.next = head;
        
         // null case
        if(head == null || head.next == null)
            return null;
        
        ListNode slow = dummy;
        ListNode fast = dummy;
        Boolean flag = false;
        while(fast !=  null && fast.next != null)
        {
            slow = slow.next;
            fast = fast.next.next;
            
            if(fast == slow)
            {flag = true; break;}
        }
        
       
        
        if(flag == true )
        { 
             slow = dummy;
            while(fast!= slow)
            {
                slow = slow.next;
                fast = fast.next;
            }
        } 
        else 
            return null;
        
        return slow;
    }
}

File: detect_cycle_in_linked_list.py
'''
Implementation: Using slow and fast pointers
Time complexity: O(n)
Space complexity: O(1)
'''
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def detectCycle(self, head: ListNode) -> ListNode:
        slow_p = head
        fast_p = head
        dummy = head
        
        while slow_p and fast_p and fast_p.next:
            slow_p = slow_p.next
            fast_p = fast_p.next.next
            if slow_p == fast_p:
                while dummy != slow_p:
                    slow_p = slow_p.next
                    dummy = dummy.next
                return slow_p
        return None